ipa-core/src/protocol/ipa_prf/aggregation/breakdown_reveal.rs

```diff
@@ -0,0 +1,423 @@
+    let atributions = shuffle_attributions::<_, _, B>(&ctx, atributions).await?;
+    let grouped_tvs = reveal_breakdowns::<HV, _, _, B>(&ctx, atributions).await?;
+    add_tvs_by_bk::<TV, HV, B>(&ctx, grouped_tvs).await
+}
+
+/// Shuffles attribution Breakdown key and Trigger Value secret shares. Input
+/// and output are the same type.
+///
+/// TODO: Use a smaller BA type to contain BK and TV
+/// TODO: Sharded shuffle
+async fn shuffle_attributions<BK, TV, const B: usize>(
+    parent_ctx: &UpgradedSemiHonestContext<'_, NotSharded, Boolean>,
+    contribs: Vec<SecretSharedAttributionOutputs<BK, TV>>,
+) -> Result<Vec<SecretSharedAttributionOutputs<BK, TV>>, Error>
+where
+    BK: BreakdownKey<B>,
+    TV: BooleanArray + U128Conversions,
+{
+    let shuffle_ctx: UpgradedSemiHonestContext<'_, NotSharded, Boolean> =
+        parent_ctx.narrow(&AggregationStep::Shuffle);
+    shuffle_attribution_outputs::<_, BK, TV, BA64>(shuffle_ctx, contribs).await
+}
+
+/// Transforms the Breakdown key from a secret share into a revealed `usize`.
+/// The input are the Atrributions and the output is a list of lists of secret
+/// shared Trigger Values. Since Breakdown Keys are assumed to be dense the
+/// first list contains all the possible Breakdowns, the index in the list
+/// representing the Breakdown value. The second list groups all the Trigger
+/// Values for that particular Breakdown.
+///
+/// TODO: Vectorize
+/// ```rust, ignore
+/// [Replicated<Boolean>;2].transpose() -> Replicated<Boolean, N=2>
+/// ```
+#[tracing::instrument(name = "reveal_breakdowns", skip_all, fields(
+    total = attributions.len(),
+))]
+async fn reveal_breakdowns<HV, BK, TV, const B: usize>(
+    parent_ctx: &UpgradedSemiHonestContext<'_, NotSharded, Boolean>,
+    attributions: Vec<SecretSharedAttributionOutputs<BK, TV>>,
+) -> Result<GroupedTriggerValues<HV, B>, Error>
+where
+    BK: BreakdownKey<B>,
+    TV: BooleanArray + U128Conversions,
+    HV: BooleanArray,
+{
+    let reveal_ctx = parent_ctx
+        .narrow(&AggregationStep::RevealStep)
+        .set_total_records(TotalRecords::specified(attributions.len())?);
+
+    let reveal_work = stream::iter(attributions).enumerate().map(|(i, ao)| {
+        let record_id = RecordId::from(i);
+        let reveal_ctx = reveal_ctx.clone();
+        async move {
+            let revealed_bk =
+                reveal(reveal_ctx, record_id, &ao.attributed_breakdown_key_bits).await?;
+            let revealed_bk: BK = BK::from_array(&revealed_bk);
+            let Ok(bk) = usize::try_from(revealed_bk.as_u128()) else {
+                return Err(Error::Internal);
+            };
+            Ok::<(usize, Replicated<TV>), Error>((bk, ao.capped_attributed_trigger_value))
+        }
+    });
+    let tv_size: usize = TV::BITS.try_into().unwrap();
+    let mut grouped_tvs = GroupedTriggerValues::<HV, B>::new(tv_size);
+    let tvs: Vec<(usize, Replicated<TV>)> = seq_join(reveal_ctx.active_work(), reveal_work)
+        .try_collect()
```

Comment:
the reason I was doing it is because I didn't wanted to modify `grouped_tvs` in an async scope. I wanted all the modifications of `grouped_tvs` to happen in the main thread. If I do something like:
```
seq_join(reveal_ctx.active_work(), reveal_work)
        .try_for_each(move |(bk, tv)| {
            grouped_tvs.push(bk, tv.to_bits())
        })
        .await?;
```
I'm getting an issue with captured variables escaping the closure body

Reply:
it works if you pin the stream and try_next it